metaflow/package/mfenv.py

```diff
@@ -0,0 +1,867 @@
+from ..extension_support import EXT_EXCLUDE_SUFFIXES, metadata, package_mfext_all
+
+from ..meta_files import (
+    MFCONF_DIR,
+    MFENV_DIR,
+    MFENV_MARKER,
+    MetaFile,
+    generic_get_filename,
+    v1_get_filename,
+    meta_file_name,
+)
+from ..util import get_metaflow_root
+
+packages_distributions = None
+
+if sys.version_info[:2] >= (3, 10):
+    packages_distributions = metadata.packages_distributions
+else:
+    # This is the code present in 3.10+ -- we replicate here for other versions
+    def _packages_distributions() -> Mapping[str, List[str]]:
+        """
+        Return a mapping of top-level packages to their
+        distributions.
+        """
+        pkg_to_dist = defaultdict(list)
+        for dist in metadata.distributions():
+            for pkg in _top_level_declared(dist) or _top_level_inferred(dist):
+                pkg_to_dist[pkg].append(dist.metadata["Name"])
+        return dict(pkg_to_dist)
+
+    def _top_level_declared(dist: metadata.Distribution) -> List[str]:
+        return (dist.read_text("top_level.txt") or "").split()
+
+    def _topmost(name: "pathlib.PurePosixPath") -> Optional[str]:
+        """
+        Return the top-most parent as long as there is a parent.
+        """
+        top, *rest = name.parts
+        return top if rest else None
+
+    def _get_toplevel_name(name: "pathlib.PurePosixPath") -> str:
+        return _topmost(name) or (
+            # python/typeshed#10328
+            inspect.getmodulename(name)  # type: ignore
+            or str(name)
+        )
+
+    def _top_level_inferred(dist: "metadata.Distribution"):
+        opt_names = set(map(_get_toplevel_name, dist.files or []))
+
+        def importable_name(name):
+            return "." not in name
+
+        return filter(importable_name, opt_names)
+
+    packages_distributions = _packages_distributions
+
+
+if TYPE_CHECKING:
+    import pathlib
+
+
+_cached_distributions = None
+
+name_normalizer = re.compile(r"[-_.]+")
+
+
+class AddToPackageType(IntEnum):
+    CODE_ONLY = 0x100
+    CONFIG_ONLY = 0x200
+    FILES_ONLY = 0x400
+    ALL = 0xFFF
+    CODE_FILE = 0x501
+    CODE_MODULE = 0x502
+    CODE_METAFLOW = 0x504
+    CONFIG_FILE = 0x601
+    CONFIG_CONTENT = 0x202
+
+
+def modules_to_distributions() -> Dict[str, List[metadata.Distribution]]:
+    """
+    Return a mapping of top-level modules to their distributions.
+
+    Returns
+    -------
+    Dict[str, List[metadata.Distribution]]
+        A mapping of top-level modules to their distributions.
+    """
+    global _cached_distributions
+    if _cached_distributions is None:
+        _cached_distributions = {
+            k: [metadata.distribution(d) for d in v]
+            for k, v in packages_distributions().items()
+        }
+    return _cached_distributions
+
+
+_ModuleInfo = NamedTuple(
+    "_ModuleInfo", [("name", str), ("root_paths", Set[str]), ("module", ModuleType)]
+)
+
+
+class PackagedDistribution(metadata.Distribution):
```

Comment:
consider moving this to a different file so that we can avoid an import from .mfnv in cli.py